### DISCUSSION TITLE
Passing a filtered array to `each`?

### DISCUSSION THREAD

**Discussion by pdaoust** [sunng87/handlebars-rust#608] [Q&A]
Hi, I'm working on someone else's code, and the context passed into one template (let's call it a `Bar`) contains one property `foos` which is a `Vec<Foo>`. Each `Foo` has a property `widget` which is used to filter in some `{{#each}}` blocks but not others. It'd be no problem if the template were simply generating content like this: `{{#each foos}}{{#if widget}}<p>{{name}}</p>{{/if}}{{/each}}`, but in some cases it's used to create a comma-separated list:

```handlebars
{{#each foos}}{{#if widget}}{{name}}{{#unless @last}}, {{/unless}}{{/if}}{{/each}}
```

As you can see, if the second-to-last `foo` is a `widget`, but the last one is not, there'll be a trailing comma, which causes the generated file to be invalid (it's generating JS code).

I'd like to avoid creating a new `foos_that_are_widgets` field on the context, so I'm wondering if it's possible to either:

* define a `foos_that_are_widgets()` function in the `impl` for `Bar` that returns the filtered vec
* filter `foos` during or before `{{#each}}`

**Comment by sunng87** [comment]
hi @pdaoust, currently it's not possible to use function as data field directly in handlebars-rust. Because internally we transformed all data into serde_json's `Value` type so we don't call any function of the data structure. 

I will suggest you to filter `foo` before rendering the template

**Comment by pdaoust** [comment]
Hi, thanks for the response. I ended up creating a `filter` helper which serves my needs -- takes a value and a filter-predicate-in-a-string, and renders a very tiny template that checks the truthiness of the predicate for each item in the array. I don't know Rust, but I managed to cargo-cult my way past `rustc`'s complaints about ownership and create a working helper!
